File: prime_calculator_cpp/prime_calculator.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <chrono>
#include <omp.h>

// Function to generate prime numbers using the Sieve of Eratosthenes
void segmentedSieve(int limit, std::vector<int> &primes) {
    int sqrtLimit = std::sqrt(limit);
    std::vector<bool> isPrime(sqrtLimit + 1, true);

    for (int i = 2; i * i <= sqrtLimit; ++i) {
        if (isPrime[i]) {
            for (int j = i * i; j <= sqrtLimit; j += i) {
                isPrime[j] = false;
            }
        }
    }

    std::vector<int> smallPrimes;
    for (int i = 2; i <= sqrtLimit; ++i) {
        if (isPrime[i]) {
            smallPrimes.push_back(i);
        }
    }

    int segmentSize = sqrtLimit;
    std::vector<bool> segment(segmentSize);

    #pragma omp parallel for schedule(dynamic)
    for (int low = sqrtLimit + 1; low <= limit; low += segmentSize) {
        std::fill(segment.begin(), segment.end(), true);

        int high = std::min(low + segmentSize - 1, limit);

        for (int prime : smallPrimes) {
            int start = std::max(prime * prime, (low + prime - 1) / prime * prime);
            for (int j = start; j <= high; j += prime) {
                segment[j - low] = false;
            }
        }

        #pragma omp critical
        {
            for (int i = low; i <= high; ++i) {
                if (segment[i - low]) {
                    primes.push_back(i);
                }
            }
        }
    }

    primes.insert(primes.begin(), smallPrimes.begin(), smallPrimes.end());
}

int main() {
    const int TARGET_PRIMES = 1000000;
    const int LIMIT = 15485863; // Approximate upper bound to ensure 1,000,000 primes
    std::vector<int> primes;

    auto start = std::chrono::high_resolution_clock::now();

    segmentedSieve(LIMIT, primes);

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    std::cout << "Elapsed time: " << elapsed.count() << " seconds" << std::endl;

    if (primes.size() < TARGET_PRIMES) {
        std::cerr << "Error: Not enough primes found. Increase the limit." << std::endl;
        return 1;
    }

    // Write the first 1,000,000 prime numbers to a file
    std::ofstream outfile("primes.txt");
    for (int i = 0; i < TARGET_PRIMES; ++i) {
        outfile << i + 1 << ". " << primes[i] << std::endl;
    }
    outfile.close();

    return 0;
}

File: prime_calculator_python/prime_calculator.py
import math
import numpy as np
from multiprocessing import Pool

def small_primes_sieve(limit):
    is_prime = np.ones(limit + 1, dtype=bool)
    is_prime[:2] = False
    for i in range(2, int(math.sqrt(limit)) + 1):
        if is_prime[i]:
            is_prime[i*i:limit+1:i] = False
    return np.nonzero(is_prime)[0]

def sieve_segment(low, high, small_primes):
    segment = np.ones(high - low + 1, dtype=bool)
    for prime in small_primes:
        start = max(prime*prime, (low + prime - 1) // prime * prime)
        if start > high:
            break
        segment[start - low:high - low + 1:prime] = False
    return np.nonzero(segment)[0] + low

def find_primes(limit, segment_size=1000000):
    sqrt_limit = int(math.sqrt(limit))
    small_primes = small_primes_sieve(sqrt_limit)
    primes = small_primes.tolist()

    segments = [(low, min(low + segment_size - 1, limit), small_primes) for low in range(sqrt_limit + 1, limit + 1, segment_size)]
    
    with Pool() as pool:
        for result in pool.starmap(sieve_segment, segments):
            primes.extend(result)

    return primes

def main():
    TARGET_PRIMES = 1000000
    LIMIT = 15485863

    primes = find_primes(LIMIT)

    if len(primes) < TARGET_PRIMES:
        print("Error: Not enough primes found. Increase the limit.")
        return

    # Write the first 1,000,000 prime numbers to a file
    with open("primes.txt", "w") as outfile:
        for i in range(TARGET_PRIMES):
            outfile.write(f"{i + 1}. {primes[i]}\n")

if __name__ == "__main__":
    main()
